src/rust/engine/src/externs/interface.rs

```diff
@@ -1399,6 +1400,48 @@ fn rule_graph_consumed_types<'py>(
     })
 }
 
+#[pyfunction]
+fn rule_graph_rule_gets<'p>(py: Python<'p>, py_scheduler: &PyScheduler) -> PyO3Result<&'p PyDict> {
+    let core = &py_scheduler.0.core;
+    core.executor.enter(|| {
+        let result = PyDict::new(py);
+        for (rule, rule_dependencies) in core.rule_graph.rule_dependencies() {
+            let Rule::Task(task) = rule else { continue };
+
+            let function = &task.func;
+            let mut dependencies = Vec::new();
+            for (dependency_key, rule) in rule_dependencies {
+                // NB: We are only migrating non-union Gets, which are those in the `gets` list
+                // which do not have `in_scope_params` marking them as being for unions, or a call
+                // signature marking them as already being call-by-name.
+                if dependency_key.call_signature.is_some()
+                    || dependency_key.in_scope_params.is_some()
+                    || !task.gets.contains(dependency_key)
+                {
+                    continue;
+                }
+                let Rule::Task(task) = rule else { continue };
```

Comment:
re-use of `task`, shadows outer scoped `task`.

Reply:
That's fairly idiomatic in Rust, apparently.

Reply:
I generally hate shadowing, but at least Rust helps prevent anything TOO egregious... Though it was a huge surprise on day 1 of learning Rust for me 🤦🏽 

https://doc.rust-lang.org/book/ch03-01-variables-and-mutability.html#shadowing

https://internals.rust-lang.org/t/history-of-shadowing-in-rust/6441/4

Shadowing in new scopes (e.g. for checking null or something) I'm pretty cool with (like in Swift)

```swift
let maybeVal: String? = "hello"
if let maybeVal {
    // Use `maybeVal` as a non-optional String 
}
```